src/java/org/apache/cassandra/service/snapshot/TableSnapshot.java

```diff
@@ -287,66 +473,79 @@ protected static String buildSnapshotId(String keyspaceName, String tableName, U
         return String.format("%s:%s:%s:%s", keyspaceName, tableName, tableId, tag);
     }
 
-    public static class SnapshotTrueSizeCalculator extends DirectorySizeCalculator
+    public static Set<Descriptor> getSnapshotDescriptors(String keyspace, String table, String tag)
     {
-        /**
-         * Snapshots are composed of hard-linked sstables. The true snapshot size should only include
-         * snapshot files which do not contain a corresponding "live" sstable file.
-         */
-        @Override
-        public boolean isAcceptable(Path snapshotFilePath)
+        try
         {
-            return !getLiveFileFromSnapshotFile(snapshotFilePath).exists();
-        }
-    }
+            Refs<SSTableReader> snapshotSSTableReaders = getSnapshotSSTableReaders(keyspace, table, tag);
 
-    /**
-     * Returns the corresponding live file for a given snapshot file.
-     *
-     * Example:
-     *  - Base table:
-     *    - Snapshot file: ~/.ccm/test/node1/data0/test_ks/tbl-e03faca0813211eca100c705ea09b5ef/snapshots/1643481737850/me-1-big-Data.db
-     *    - Live file: ~/.ccm/test/node1/data0/test_ks/tbl-e03faca0813211eca100c705ea09b5ef/me-1-big-Data.db
-     *  - Secondary index:
-     *    - Snapshot file: ~/.ccm/test/node1/data0/test_ks/tbl-e03faca0813211eca100c705ea09b5ef/snapshots/1643481737850/.tbl_val_idx/me-1-big-Summary.db
-     *    - Live file: ~/.ccm/test/node1/data0/test_ks/tbl-e03faca0813211eca100c705ea09b5ef/.tbl_val_idx/me-1-big-Summary.db
-     *
-     */
-    static File getLiveFileFromSnapshotFile(Path snapshotFilePath)
-    {
-        // Snapshot directory structure format is {data_dir}/snapshots/{snapshot_name}/{snapshot_file}
-        Path liveDir = snapshotFilePath.getParent().getParent().getParent();
-        if (Directories.isSecondaryIndexFolder(snapshotFilePath.getParent()))
+            Set<Descriptor> descriptors = new HashSet<>();
+            for (SSTableReader ssTableReader : snapshotSSTableReaders)
+            {
+                descriptors.add(ssTableReader.descriptor);
+            }
+
+            return descriptors;
+        }
+        catch (IOException e)
         {
-            // Snapshot file structure format is {data_dir}/snapshots/{snapshot_name}/.{index}/{sstable-component}.db
-            liveDir = File.getPath(liveDir.getParent().toString(), snapshotFilePath.getParent().getFileName().toString());
+            throw Throwables.unchecked(e);
         }
-        return new File(liveDir.toString(), snapshotFilePath.getFileName().toString());
     }
 
-    public static Predicate<TableSnapshot> shouldClearSnapshot(String tag, long olderThanTimestamp)
+    public static Refs<SSTableReader> getSnapshotSSTableReaders(String keyspace, String table, String tag) throws IOException
```

Comment:
I'd like to have a signature here that takes a CFS, any reason we don't have it? it can simplify the call sites
```suggestion
    public static Refs<SSTableReader> getSnapshotSSTableReaders(String keyspace, String table, String tag) throws IOException
    {
        return getSnapshotSSTableReaders(Keyspace.open(keyspace).getColumnFamilyStore(table), tag);
    }

    public static Refs<SSTableReader> getSnapshotSSTableReaders(ColumnFamilyStore cfs, String tag)
```

Reply:
will be done